exonum/src/runtime/dispatcher/tests.rs

```diff
@@ -568,3 +573,330 @@ fn test_shutdown() {
+                    Ok(())
+                } else {
+                    Err(ExecutionError::new(error_kind, "bad artifact!"))
+                }
+            }
+            _ => unreachable!(),
+        };
+
+        let artifacts = Arc::clone(&self.artifacts);
+        let task = future::lazy(move || {
+            // This isn't correct way to delay future completion, but the correct way
+            // (`tokio::timer::Delay`) cannot be used since the futures
+            // returned by `Runtime::deploy_artifact()`are not (yet?) run
+            // on the `tokio` runtime.
+            thread::sleep(delay);
+            result
+        })
+        .then(move |res| {
+            let mut artifacts = artifacts.lock().unwrap();
+            let status = artifacts.entry(artifact.name).or_default();
+            status.attempts += 1;
+            if res.is_ok() {
+                status.is_deployed = true;
+            }
+            res
+        });
+        Box::new(task)
+    }
+
+    fn is_artifact_deployed(&self, id: &ArtifactId) -> bool {
+        self.artifacts
+            .lock()
+            .unwrap()
+            .get(&id.name)
+            .copied()
+            .unwrap_or_default()
+            .is_deployed
+    }
+
+    fn start_adding_service(
+        &self,
+        _context: ExecutionContext<'_>,
+        _spec: &InstanceSpec,
+        _parameters: Vec<u8>,
+    ) -> Result<(), ExecutionError> {
+        Ok(())
+    }
+
+    fn commit_service(
+        &mut self,
+        _snapshot: &dyn Snapshot,
+        _spec: &InstanceSpec,
+    ) -> Result<(), ExecutionError> {
+        Ok(())
+    }
+
+    fn execute(
+        &self,
+        _context: ExecutionContext<'_>,
+        _call_info: &CallInfo,
+        _parameters: &[u8],
+    ) -> Result<(), ExecutionError> {
+        Ok(())
+    }
+
+    fn state_hashes(&self, _snapshot: &dyn Snapshot) -> StateHashAggregator {
+        StateHashAggregator::default()
+    }
+
+    fn before_commit(
+        &self,
+        _context: ExecutionContext<'_>,
+        _id: InstanceId,
+    ) -> Result<(), ExecutionError> {
+        Ok(())
+    }
+
+    fn after_commit(&mut self, _snapshot: &dyn Snapshot, mailbox: &mut Mailbox) {
+        for action in mem::replace(&mut *self.mailbox_actions.lock().unwrap(), vec![]) {
+            mailbox.push(action);
+        }
+    }
+}
+
+#[test]
+fn delayed_deployment() {
+    let db = Arc::new(TemporaryDB::new());
+    let blockchain = Blockchain::new(
+        Arc::clone(&db) as Arc<dyn Database>,
+        gen_keypair(),
+        ApiSender(mpsc::channel(1).0),
+    );
+    let runtime = DeploymentRuntime::default();
+    let mut dispatcher = DispatcherBuilder::new()
+        .with_runtime(2, runtime.clone())
+        .finalize(&blockchain);
+
+    let mut fork = db.fork();
+    create_genesis_block(&mut dispatcher, &mut fork);
+    db.merge_sync(fork.into_patch()).unwrap();
+
+    // Queue an artifact for deployment.
+    let artifact = ArtifactId {
```

Comment:
Nit: code from
```rust
// Queue an artifact for deployment.
```

till

```
db.merge_sync(fork.into_patch()).unwrap();
```

is almost copy-pasted several times. Maybe it worth moving into a separate function?